File: zig_build/tests/constants/constants_parity.c
// SPDX-License-Identifier: GPL-3.0-only

#include <stdio.h>
#include <string.h>

#include "constants_test_runtime.h"

void fnConstant(uint16_t constant);
void fnPi(uint16_t unusedButMandatoryParameter);

void oracle_fnConstant(uint16_t constant);
void oracle_fnPi(uint16_t unusedButMandatoryParameter);

typedef void (*constants_fn)(uint16_t);

static int reportMismatch(const char *name,
                          uint16_t arg,
                          const constants_snapshot_t *expected,
                          const constants_snapshot_t *actual) {
  if(memcmp(expected, actual, sizeof(*expected)) == 0) {
    return 0;
  }

  fprintf(stderr,
          "%s(%u) parity mismatch\n"
          "  expected: lift=%u solver=%#x realloc=%u/%d/%u/%u/%u real=%u/%u/%d convert=%u/%u/%d/%d adjust=%u/%d/%u/%u/%d/%d/%d zero=%u x=%u info=%u/%s/%s\n"
          "  actual:   lift=%u solver=%#x realloc=%u/%d/%u/%u/%u real=%u/%u/%d convert=%u/%u/%d/%d adjust=%u/%d/%u/%u/%d/%d/%d zero=%u x=%u info=%u/%s/%s\n",
          name,
          arg,
          expected->lift_stack_calls,
          expected->current_solver_status,
          expected->reallocate_register_calls,
          expected->reallocate_register_reg,
          expected->reallocate_register_data_type,
          expected->reallocate_register_data_size_without_data_len_blocks,
          expected->reallocate_register_tag,
          expected->real_to_real34_calls,
          expected->real_to_real34_source_id,
          expected->real_to_real34_destination_reg,
          expected->convert_real_to_result_register_calls,
          expected->convert_real_to_result_source_id,
          expected->convert_real_to_result_dest,
          expected->convert_real_to_result_angle,
          expected->adjust_result_calls,
          expected->adjust_result_res,
          expected->adjust_result_drop_y,
          expected->adjust_result_set_cpx_res,
          expected->adjust_result_op1,
          expected->adjust_result_op2,
          expected->adjust_result_op3,
          expected->set_lastinteger_base_to_zero_calls,
          expected->register_x_real34_id,
          expected->more_info_calls,
          expected->more_info_line1,
          expected->more_info_line2,
          actual->lift_stack_calls,
          actual->current_solver_status,
          actual->reallocate_register_calls,
          actual->reallocate_register_reg,
          actual->reallocate_register_data_type,
          actual->reallocate_register_data_size_without_data_len_blocks,
          actual->reallocate_register_tag,
          actual->real_to_real34_calls,
          actual->real_to_real34_source_id,
          actual->real_to_real34_destination_reg,
          actual->convert_real_to_result_register_calls,
          actual->convert_real_to_result_source_id,
          actual->convert_real_to_result_dest,
          actual->convert_real_to_result_angle,
          actual->adjust_result_calls,
          actual->adjust_result_res,
          actual->adjust_result_drop_y,
          actual->adjust_result_set_cpx_res,
          actual->adjust_result_op1,
          actual->adjust_result_op2,
          actual->adjust_result_op3,
          actual->set_lastinteger_base_to_zero_calls,
          actual->register_x_real34_id,
          actual->more_info_calls,
          actual->more_info_line1,
          actual->more_info_line2);
  return 1;
}

static int runCase(const char *name,
                   constants_fn oracle_fn,
                   constants_fn zig_fn,
                   uint16_t arg) {
  constants_snapshot_t expected;
  constants_snapshot_t actual;

  constantsRuntimeReset();
  oracle_fn(arg);
  constantsRuntimeCapture(&expected);

  constantsRuntimeReset();
  zig_fn(arg);
  constantsRuntimeCapture(&actual);

  return reportMismatch(name, arg, &expected, &actual);
}

int main(void) {
  int failures = 0;

  failures += runCase("fnConstant", oracle_fnConstant, fnConstant, 0);
  failures += runCase("fnConstant", oracle_fnConstant, fnConstant, 7);
  failures += runCase("fnConstant", oracle_fnConstant, fnConstant, NOUC - 1);
  failures += runCase("fnConstant", oracle_fnConstant, fnConstant, NOUC);
  failures += runCase("fnPi", oracle_fnPi, fnPi, 0);
  failures += runCase("fnPi", oracle_fnPi, fnPi, 99);

  if(failures != 0) {
    fprintf(stderr, "%d constants parity checks failed\n", failures);
    return 1;
  }

  return 0;
}

File: zig_build/tests/constants/constants_parity.zig
const std = @import("std");

const constants_snapshot_t = extern struct {
    lift_stack_calls: u32,
    current_solver_status: u16,

    reallocate_register_calls: u32,
    reallocate_register_reg: i16,
    reallocate_register_data_type: u32,
    reallocate_register_data_size_without_data_len_blocks: u16,
    reallocate_register_tag: u32,

    real_to_real34_calls: u32,
    real_to_real34_source_id: u16,
    real_to_real34_destination_reg: i16,

    convert_real_to_result_register_calls: u32,
    convert_real_to_result_source_id: u16,
    convert_real_to_result_dest: i16,
    convert_real_to_result_angle: i32,

    adjust_result_calls: u32,
    adjust_result_res: i16,
    adjust_result_drop_y: bool,
    adjust_result_set_cpx_res: bool,
    adjust_result_op1: i16,
    adjust_result_op2: i16,
    adjust_result_op3: i16,

    set_lastinteger_base_to_zero_calls: u32,
    register_x_real34_id: u16,

    more_info_calls: u32,
    more_info_line1: [64]u8,
    more_info_line2: [256]u8,
};

extern fn constantsRuntimeReset() void;
extern fn constantsRuntimeCapture(snapshot: *constants_snapshot_t) void;
extern fn fnConstant(constant: u16) callconv(.c) void;
extern fn fnPi(unused_but_mandatory_parameter: u16) callconv(.c) void;
extern fn oracle_fnConstant(constant: u16) callconv(.c) void;
extern fn oracle_fnPi(unused_but_mandatory_parameter: u16) callconv(.c) void;

const ConstantsFn = *const fn (u16) callconv(.c) void;

fn runCase(name: []const u8, oracle_fn: ConstantsFn, zig_fn: ConstantsFn, arg: u16) bool {
    var expected: constants_snapshot_t = undefined;
    var actual: constants_snapshot_t = undefined;

    constantsRuntimeReset();
    oracle_fn(arg);
    constantsRuntimeCapture(&expected);

    constantsRuntimeReset();
    zig_fn(arg);
    constantsRuntimeCapture(&actual);

    const equal = std.mem.eql(u8, std.mem.asBytes(&expected), std.mem.asBytes(&actual));
    if (!equal) {
        std.debug.print("{s}({d}) parity mismatch\n", .{ name, arg });
    }
    return equal;
}

pub fn main() !void {
    var failures: u32 = 0;

    if (!runCase("fnConstant", &oracle_fnConstant, &fnConstant, 0)) failures += 1;
    if (!runCase("fnConstant", &oracle_fnConstant, &fnConstant, 7)) failures += 1;
    if (!runCase("fnConstant", &oracle_fnConstant, &fnConstant, 83)) failures += 1;
    if (!runCase("fnConstant", &oracle_fnConstant, &fnConstant, 84)) failures += 1;
    if (!runCase("fnPi", &oracle_fnPi, &fnPi, 0)) failures += 1;
    if (!runCase("fnPi", &oracle_fnPi, &fnPi, 99)) failures += 1;

    if (failures != 0) {
        std.debug.print("{d} constants parity checks failed\n", .{failures});
        std.process.exit(1);
    }
}
